File: src/topo_shadow_box/core/map_insert.py
"""Map insert generation: SVG background map and 3D flat plate."""

from ..state import Bounds, Colors
from .coords import GeoToModelTransform
from .models import MeshResult


def generate_map_insert_svg(
    bounds: Bounds,
    features: dict,
    gpx_tracks: list,
    colors: Colors,
) -> str:
    """Generate an SVG map of features for paper printing.

    Returns SVG string. Stored in state for later export.
    """
    # Coordinate transform: geo -> SVG viewport
    width = 800  # SVG pixels
    lat_range = bounds.lat_range
    lon_range = bounds.lon_range
    import math
    lon_scale = math.cos(math.radians(bounds.center_lat))
    aspect = (lon_range * lon_scale) / lat_range if lat_range > 0 else 1.0
    height = int(width / aspect) if aspect > 0 else width

    def geo_to_svg(lat: float, lon: float) -> tuple[float, float]:
        x = (lon - bounds.west) / lon_range * width if lon_range > 0 else 0
        y = (bounds.north - lat) / lat_range * height if lat_range > 0 else 0
        return x, y

    parts = [
        f'<svg xmlns="http://www.w3.org/2000/svg" '
        f'width="{width}" height="{height}" '
        f'viewBox="0 0 {width} {height}">',
        f'<rect width="{width}" height="{height}" fill="{colors.map_insert}"/>',
    ]

    # Water bodies
    for water in features.get("water", []):
        coords = water.get("coordinates", [])
        if len(coords) < 3:
            continue
        points = " ".join(f"{geo_to_svg(c['lat'], c['lon'])[0]:.1f},{geo_to_svg(c['lat'], c['lon'])[1]:.1f}" for c in coords)
        parts.append(f'<polygon points="{points}" fill="{colors.water}" opacity="0.6"/>')

    # Roads
    for road in features.get("roads", []):
        coords = road.get("coordinates", [])
        if len(coords) < 2:
            continue
        points = " ".join(f"{geo_to_svg(c['lat'], c['lon'])[0]:.1f},{geo_to_svg(c['lat'], c['lon'])[1]:.1f}" for c in coords)
        parts.append(f'<polyline points="{points}" fill="none" stroke="{colors.roads}" stroke-width="1" opacity="0.5"/>')

    # Buildings
    for bldg in features.get("buildings", []):
        coords = bldg.get("coordinates", [])
        if len(coords) < 3:
            continue
        points = " ".join(f"{geo_to_svg(c['lat'], c['lon'])[0]:.1f},{geo_to_svg(c['lat'], c['lon'])[1]:.1f}" for c in coords)
        parts.append(f'<polygon points="{points}" fill="{colors.buildings}" opacity="0.4"/>')

    # GPX tracks
    for track in gpx_tracks:
        pts = track.get("points", [])
        if len(pts) < 2:
            continue
        points = " ".join(f"{geo_to_svg(p['lat'], p['lon'])[0]:.1f},{geo_to_svg(p['lat'], p['lon'])[1]:.1f}" for p in pts)
        parts.append(f'<polyline points="{points}" fill="none" stroke="{colors.gpx_track}" stroke-width="2"/>')

    parts.append("</svg>")
    return "\n".join(parts)


def generate_map_insert_plate(
    bounds: Bounds,
    features: dict,
    gpx_tracks: list,
    transform: GeoToModelTransform,
    plate_thickness_mm: float = 1.0,
) -> MeshResult:
    """Generate a thin 3D plate for the map insert.

    The plate is a flat rectangle matching the model dimensions, with features
    as very slightly raised regions for visual/tactile effect.

    Returns dict with vertices and faces.
    """
    w = transform.model_width_x
    h = transform.model_width_z
    t = plate_thickness_mm

    # Place the plate below the terrain (at -base_height - some offset)
    y_top = 0.0
    y_bot = -t

    vertices = [
        [0, y_top, 0],      # 0: top NW
        [w, y_top, 0],      # 1: top NE
        [w, y_top, h],      # 2: top SE
        [0, y_top, h],      # 3: top SW
        [0, y_bot, 0],      # 4: bot NW
        [w, y_bot, 0],      # 5: bot NE
        [w, y_bot, h],      # 6: bot SE
        [0, y_bot, h],      # 7: bot SW
    ]

    faces = [
        # Top
        [0, 1, 2], [0, 2, 3],
        # Bottom
        [4, 6, 5], [4, 7, 6],
        # Front
        [3, 2, 6], [3, 6, 7],
        # Back
        [0, 5, 1], [0, 4, 5],
        # Left
        [0, 3, 7], [0, 7, 4],
        # Right
        [1, 6, 2], [1, 5, 6],
    ]

    return MeshResult(vertices=vertices, faces=faces, name="Map Insert", feature_type="map_insert")

File: go/internal/core/map_insert.go
package core

import (
	"fmt"
	"math"
	"strings"

	"github.com/huslage/topo-shadow-box/internal/session"
)

func GenerateMapInsertSVG(bounds session.Bounds, features *session.OsmFeatureSet, gpxTracks []session.GpxTrack, colors session.Colors) string {
	width := 800.0
	latRange := bounds.LatRange()
	lonRange := bounds.LonRange()
	lonScale := math.Cos(math.Pi / 180.0 * bounds.CenterLat())
	aspect := 1.0
	if latRange > 0 {
		aspect = (lonRange * lonScale) / latRange
		if aspect <= 0 {
			aspect = 1
		}
	}
	height := width / aspect
	if height < 100 {
		height = 100
	}
	if height > 2000 {
		height = 2000
	}

	geoToSVG := func(lat, lon float64) (float64, float64) {
		x := 0.0
		y := 0.0
		if lonRange > 0 {
			x = (lon - bounds.West) / lonRange * width
		}
		if latRange > 0 {
			y = (bounds.North - lat) / latRange * height
		}
		return x, y
	}

	parts := []string{
		fmt.Sprintf(`<svg xmlns="http://www.w3.org/2000/svg" width="%.0f" height="%.0f" viewBox="0 0 %.0f %.0f">`, width, height, width, height),
		fmt.Sprintf(`<rect width="%.0f" height="%.0f" fill="%s"/>`, width, height, colors.MapInsert),
	}

	if features != nil {
		for _, w := range features.Water {
			if len(w.Coordinates) < 3 {
				continue
			}
			pts := make([]string, 0, len(w.Coordinates))
			for _, c := range w.Coordinates {
				x, y := geoToSVG(c.Lat, c.Lon)
				pts = append(pts, fmt.Sprintf("%.1f,%.1f", x, y))
			}
			parts = append(parts, fmt.Sprintf(`<polygon points="%s" fill="%s" opacity="0.6"/>`, strings.Join(pts, " "), colors.Water))
		}

		for _, r := range features.Roads {
			if len(r.Coordinates) < 2 {
				continue
			}
			pts := make([]string, 0, len(r.Coordinates))
			for _, c := range r.Coordinates {
				x, y := geoToSVG(c.Lat, c.Lon)
				pts = append(pts, fmt.Sprintf("%.1f,%.1f", x, y))
			}
			parts = append(parts, fmt.Sprintf(`<polyline points="%s" fill="none" stroke="%s" stroke-width="1" opacity="0.5"/>`, strings.Join(pts, " "), colors.Roads))
		}

		for _, b := range features.Buildings {
			if len(b.Coordinates) < 3 {
				continue
			}
			pts := make([]string, 0, len(b.Coordinates))
			for _, c := range b.Coordinates {
				x, y := geoToSVG(c.Lat, c.Lon)
				pts = append(pts, fmt.Sprintf("%.1f,%.1f", x, y))
			}
			parts = append(parts, fmt.Sprintf(`<polygon points="%s" fill="%s" opacity="0.4"/>`, strings.Join(pts, " "), colors.Buildings))
		}
	}

	for _, tr := range gpxTracks {
		if len(tr.Points) < 2 {
			continue
		}
		pts := make([]string, 0, len(tr.Points))
		for _, p := range tr.Points {
			x, y := geoToSVG(p.Lat, p.Lon)
			pts = append(pts, fmt.Sprintf("%.1f,%.1f", x, y))
		}
		parts = append(parts, fmt.Sprintf(`<polyline points="%s" fill="none" stroke="%s" stroke-width="2"/>`, strings.Join(pts, " "), colors.GpxTrack))
	}

	parts = append(parts, `</svg>`)
	return strings.Join(parts, "\n")
}

func GenerateMapInsertPlate(bounds session.Bounds, transform GeoToModelTransform, plateThicknessMM float64) *session.Mesh {
	_ = bounds
	if plateThicknessMM <= 0 {
		plateThicknessMM = 1.0
	}
	w := transform.ModelWidthX
	h := transform.ModelWidthZ
	t := plateThicknessMM

	yTop := 0.0
	yBot := -t
	vertices := [][3]float64{
		{0, yTop, 0}, {w, yTop, 0}, {w, yTop, h}, {0, yTop, h},
		{0, yBot, 0}, {w, yBot, 0}, {w, yBot, h}, {0, yBot, h},
	}
	faces := [][3]int{
		{0, 1, 2}, {0, 2, 3},
		{4, 6, 5}, {4, 7, 6},
		{3, 2, 6}, {3, 6, 7},
		{0, 5, 1}, {0, 4, 5},
		{0, 3, 7}, {0, 7, 4},
		{1, 6, 2}, {1, 5, 6},
	}
	return &session.Mesh{Vertices: vertices, Faces: faces, Name: "Map Insert", FeatureType: "map_insert"}
}
